llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -7491,3 +7491,27 @@ bool CombinerHelper::matchTruncInteger(const MachineInstr &MI,
 
   return true;
 }
+
+bool CombinerHelper::matchAnyextInteger(const MachineInstr &MI,
+                                        APInt &MatchInfo) {
+  const GAnyExt *Anyext = cast<GAnyExt>(&MI);
+
+  std::optional<APInt> Input = getIConstantVRegVal(Anyext->getSrcReg(), MRI);
+  if (!Input)
+    return false;
+
+  LLT DstTy = MRI.getType(Anyext->getReg(0));
+  LLT SrcTy = MRI.getType(Anyext->getSrcReg());
+  const auto &TLI = getTargetLowering();
+
+  if (!isConstantLegalOrBeforeLegalizer(DstTy))
+    return false;
+
+  // Some targets like RISCV prefer to sign extend some types.
+  if (TLI.isSExtCheaperThanZExt(getMVTForLLT(SrcTy), getMVTForLLT(DstTy)))
```

Comment:
I don't think this profitability check makes much sense. It would make more sense to check if the resulting immediate is legal, and if it's not, check if it is legal in the other extended case.

If you're going to use it, add an LLT overload and don't have the LLT->MVT conversion be done here.

Reply:
See
https://github.com/llvm/llvm-project/blob/cc4ec6daf0d4f43110e8220d542c1155b8c1ef51/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L6375

and 

https://github.com/llvm/llvm-project/blob/efab4a380f36dcd23561633f8bba484036c500f3/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L7432

There is a legality check for DstTy above.

Reply:
It is not a profitability check. It is about that RISC-V picks sext for anyext and AArch64 picks zext. The result may differ.

Reply:
But the cast isn't actually being performed. I don't think the DAG code makes any more sense.